// multi-integer-range/lib/esm/MultiRange.js
import * as mr from './fp.js';
/**
 * Parses and manipulates multiple integer ranges.
 * This class exists for compatibility purposes.
 * Prefer the function style API instead.
 */
export class MultiRange {
    /**
     * Creates a new MultiRange object.
     */
    constructor(data, options) {
        this.ranges = [];
        this.options = {
            parseNegative: !!(options || {}).parseNegative,
            parseUnbounded: !!(options || {}).parseUnbounded
        };
        if (typeof data === 'string') {
            this.ranges = mr.parse(data, options);
        }
        else if (typeof data === 'number' || Array.isArray(data)) {
            this.ranges = mr.normalize(data);
        }
        else if (data instanceof MultiRange) {
            this.ranges = data.ranges;
            if (options === undefined)
                this.options = data.options;
        }
        else if (data !== undefined) {
            throw new TypeError('Invalid input');
        }
    }
    /**
     * Clones this instance.
     * @returns The cloned instance.
     */
    clone() {
        return new MultiRange(this);
    }
    /**
     * Appends to this instance.
     * @param value - The data to append.
     */
    append(value) {
        this.ranges = mr.append(this.ranges, new MultiRange(value, this.options).ranges);
        return this;
    }
    /**
     * Subtracts from this instance.
     * @param value - The data to subtract.
     */
    subtract(value) {
        this.ranges = mr.subtract(this.ranges, new MultiRange(value, this.options).ranges);
        return this;
    }
    /**
     * Remove integers which are not included in `value`,
     * yielding the intersection of this and `value`.
     * @param value - The data to calculate the intersetion.
     */
    intersect(value) {
        this.ranges = mr.intersect(this.ranges, new MultiRange(value, this.options).ranges);
        return this;
    }
    /**
     * Exports the whole range data as an array of arrays.
     * @returns An copied array of range segments.
     */
    getRanges() {
        const result = [];
        for (let r of this.ranges)
            result.push([r[0], r[1]]);
        return result;
    }
    /**
     * Checks if this instance contains the specified value.
     * @param value - Value to be checked.
     * @returns True if the specified value is included in the instance.
     */
    has(value) {
        if (value === undefined)
            throw new TypeError('Invalid input');
        return mr.has(this.ranges, new MultiRange(value, this.options).ranges);
    }
    /**
     * Returns the number of range segments.
     * For example, the segmentLength of `2-5,7,9-11` is 3.
     * @returns The number of segments. Returns 0 for an empty instance.
     */
    segmentLength() {
        return this.ranges.length;
    }
    /**
     * Calculates how many numbers are effectively included in this instance.
     * For example, the length of `1-10,51-60,90` is 21.
     * @returns The number of integer values in this instance.
     *    Returns `Infinity` for an unbounded range.
     */
    length() {
        return mr.length(this.ranges);
    }
    /**
     * Checks if two instances of MultiRange are identical.
     * @param cmp - The data to compare.
     * @returns True if `cmp` is exactly the same as this instance.
     */
    equals(cmp) {
        if (cmp === undefined)
            throw new TypeError('Invalid input');
        return mr.equals(this.ranges, new MultiRange(cmp, this.options).ranges);
    }
    /**
     * Checks if the current instance is unbounded (i.e., infinite).
     */
    isUnbounded() {
        return mr.isUnbounded(this.ranges);
    }
    /**
     * Returns the minimum integer contained in this insntance.
     * Can be -Infinity or undefined.
     * @returns The minimum integer of this instance.
     */
    min() {
        return mr.min(this.ranges);
    }
    /**
     * Returns the maximum number contained in this insntance.
     * Can be Infinity or undefined.
     * @returns The maximum integer of this instance.
     */
    max() {
        return mr.max(this.ranges);
    }
    /**
     * Removes the smallest integer from this instance and returns it.
     * @returns The minimum integer removed from this instance.
     */
    shift() {
        const min = this.min();
        this.ranges = mr.tail(this.ranges);
        return min;
    }
    /**
     * Removes the largest integer from this instance and returns it.
     * @returns The maximum integer removed from this instance.
     */
    pop() {
        const max = this.max();
        this.ranges = mr.init(this.ranges);
        return max;
    }
    /**
     * Returns the string respresentation of this MultiRange.
     */
    toString() {
        return mr.stringify(this.ranges);
    }
    /**
     * Builds a flat array of integers which holds all elements in this instance.
     * Note that this may be slow and memory-consuming for large ranges.
     * Consider using the iterator whenever possible.
     */
    toArray() {
        return mr.flatten(this.ranges);
    }
    /**
     * Returns an ES6-compatible iterator.
     */
    getIterator() {
        return mr.iterate(this.ranges)[Symbol.iterator]();
    }
    [Symbol.iterator]() {
        return mr.iterate(this.ranges)[Symbol.iterator]();
    }
}
export const multirange = (data, options) => new MultiRange(data, options);
